### ISSUE TITLE
bug report: thread 'main' has overflowed its stack

### ISSUE THREAD

**Issue by vootooguru** [trailbaseio/trailbase#16]
This is a fantastic project that fully leverages the advantages of Rust. I hope to port the app I am currently developing to Trailbase and extend its functionality. However, when running example/custom-binary, the program throws an error: thread 'main' has overflowed its stack. This seems like a bug, and I currently don't have a solution for it. Can anyone help me with this?

log info: 
<details><summary>Details</summary>
<p>

Running `D:\Github\trailbase\target\debug\custom-binary.exe`
[2024-12-25T08:09:40Z INFO  trailbase_core::data_dir] Initialized fresh data dir: "./traildepot/"
[2024-12-25T08:09:40Z INFO  trailbase_core::migrations] Logs migrations: Report { applied_migrations: [Migration { state: Applied, name: "initial", checksum: 3314206368411190683, version: 1, prefix: Versioned, sql: Some("CREATE TABLE IF NOT EXISTS _logs (\r\n  -- NOTE: We're skipping the CHECK(is_uuid_v7) here as mostly inconsequential\r\n  -- micro-optimization but we also don't wanna expose the logs via RecordAPIs,\r\n  -- so there's not strict need.\r\n  id                           BLOB PRIMARY KEY DEFAULT (uuid_v7()) NOT NULL,\r\n  -- Timestamp in seconds with fractional millisecond resolution.\r\n  created                      REAL DEFAULT (UNIXEPOCH('subsec')) NOT NULL,\r\n  -- Entry type. We could probably also split by table :shrug:\r\n  type                      
   INTEGER DEFAULT 0 NOT NULL,\r\n\r\n  level                        INTEGER DEFAULT 0 NOT NULL,\r\n  status                       INTEGER DEFAULT 0 NOT NULL,\r\n  method                       TEXT DEFAULT '' NOT NULL,\r\n  url                          TEXT DEFAULT '' NOT NULL,\r\n\r\n  latency                      REAL DEFAULT 0 NOT NULL,\r\n\r\n  client_ip                    TEXT DEFAULT '' NOT NULL,\r\n  referer                      TEXT DEFAULT '' NOT NULL,\r\n  user_agent                   TEXT DEFAULT '' NOT NULL,\r\n\r\n  data                         BLOB\r\n) strict;\r\n\r\nCREATE INDEX IF NOT EXISTS __logs__level_index ON _logs (level);\r\nCREATE INDEX IF NOT EXISTS __logs__created_index ON _logs (created);\r\nCREATE INDEX IF NOT EXISTS __logs__status_index ON _logs (status);\r\nCREATE INDEX IF NOT EXISTS __logs__method_index ON _logs (method);\r\n"), applied_on: Some(2024-12-25 8:09:40.2171011 +00:00:00) }] }
[2024-12-25T08:09:40Z INFO  trailbase_core::migrations] applied migration: Migration { state: Applied, name: "initial", checksum: 13644041026481178259, version: 1, prefix: Versioned, sql: Some("--\r\n-- User table.\r\n--\r\nCREATE TABLE _user (\r\n  id         
                      BLOB PRIMARY KEY NOT NULL CHECK(is_uuid_v7(id)) DEFAULT (uuid_v7()),\r\n  email                            TEXT NOT NULL CHECK(is_email(email)),\r\n  password_hash                    TEXT DEFAULT '' NOT NULL,\r\n  verified                 
        INTEGER DEFAULT FALSE NOT NULL,\r\n  admin                            INTEGER DEFAULT FALSE NOT NULL,\r\n\r\n  created                          INTEGER DEFAULT (UNIXEPOCH()) NOT NULL,\r\n  updated                          INTEGER DEFAULT (UNIXEPOCH()) NOT NULL,\r\n\r\n  -- Ephemeral data for auth flows.\r\n  --\r\n  -- Email change/verification flow.\r\n  email_verification_code          TEXT,\r\n  email_verification_code_sent_at  INTEGER,\r\n  -- Change email flow.\r\n  pending_email                    TEXT CHECK(is_email(pending_email)),\r\n  -- Reset forgotten password flow.\r\n  password_reset_code              TEXT,\r\n  password_reset_code_sent_at      INTEGER,\r\n  -- Authorization Code Flow (optionally with PKCE proof key).\r\n  authorization_code          
     TEXT,\r\n  authorization_code_sent_at       INTEGER,\r\n  pkce_code_challenge              TEXT,\r\n\r\n  -- OAuth metadata\r\n  --\r\n  -- provider_id maps to proto.config.OAuthProviderId enum.\r\n  provider_id                      INTEGER DEFAULT 0 NOT NULL,\r\n  -- The external provider's id for the user.\r\n  provider_user_id                 TEXT,\r\n  -- Link to an external avatar image for oauth providers only.\r\n  provider_avatar_url              TEXT\r\n) STRICT;\r\n\r\nCREATE UNIQUE INDEX __user__email_index ON _user (email);\r\nCREATE UNIQUE INDEX __user__email_verification_code_index ON _user (email_verification_code);\r\nCREATE UNIQUE INDEX __user__password_reset_code_index ON _user (password_reset_code);\r\nCREATE UNIQUE INDEX __user__authorization_code_index ON _user (authorization_code);\r\nCREATE UNIQUE INDEX __user__provider_ids_index ON _user (provider_id, provider_user_id);\r\n\r\nCREATE TRIGGER __user__updated_trigger AFTER UPDATE ON _user FOR EACH ROW\r\n  BEGIN\r\n    UPDATE _user SET updated = UNIXEPOCH() WHERE id = OLD.id;\r\n  END;\r\n\r\n--\r\n-- Session table\r\n--\r\nCREATE TABLE _session (\r\n  id                           INTEGER PRIMARY KEY NOT NULL,\r\n  user                         BLOB NOT NULL REFERENCES _user(id) ON DELETE CASCADE,\r\n  refresh_token                TEXT NOT NULL,\r\n  updated                      INTEGER DEFAULT (UNIXEPOCH()) NOT NULL\r\n) STRICT;\r\n\r\n-- NOTE: The expiry is computed based on `updated` + TTL, thus touching the row\r\n-- will extend the opaque refresh token's expiry.\r\nCREATE TRIGGER __session__updated_trigger AFTER UPDATE ON _session FOR EACH ROW\r\n  BEGIN\r\n    UPDATE _session SET updated = UNIXEPOCH() WHERE user = OLD.user;\r\n  END;\r\n\r\n-- Main unique index to lookup refresh tokens efficiently.\r\nCREATE UNIQUE INDEX __session__refresh_token_index ON _session (refresh_token);\r\n-- An index on the user for efficient deletions of all sessions given a user.\r\nCREATE INDEX __session__user_index ON _session (user);\r\n\r\n--\r\n-- User avatar table\r\n--\r\nCREATE TABLE _user_avatar (\r\n  user                         BLOB PRIMARY KEY NOT NULL REFERENCES _user(id) ON DELETE CASCADE,\r\n  file                         TEXT CHECK(jsonschema('std.FileUpload', file, 'image/png, image/jpeg')) NOT NULL,\r\n  updated                 
     INTEGER DEFAULT (UNIXEPOCH()) NOT NULL\r\n) STRICT;\r\n\r\nCREATE TRIGGER __user_avatar__updated_trigger AFTER UPDATE ON _user_avatar FOR EACH ROW\r\n  BEGIN\r\n    UPDATE _user_avatar SET updated = UNIXEPOCH() WHERE user = OLD.user;\r\n  END;\r\n"), applied_on: Some(2024-12-25 8:09:40.2465698 +00:00:00) }

thread 'main' has overflowed its stack
error: process didn't exit successfully: `D:\Github\trailbase\target\debug\custom-binary.exe` (exit code: 0xc00000fd, STATUS_STACK_OVERFLOW)

</p>
</details>

**Comment by ignatz** [comment]
Hi @vootooguru,

this came up in https://github.com/trailbaseio/trailbase/discussions/13, is further discussed in https://github.com/gwenn/lemon-rs/issues/78, and I have PR out to address it:  https://github.com/gwenn/lemon-rs/pull/79

TL;DR: Windows has a smaller default stack size of 1MB (which is still generous), however the parser implementation https://github.com/gwenn/lemon-rs in combination with how the rust compiler allocates stack space in debug builds makes it go beyond that limit. Ways to work around until it has been fixed upstream:

 * Build in release mode
 * Increase stack size (e.g. rustflags = ["-C", "link-arg=/STACK:4000000"])
 * Patch my PR

Let's hope we can address this quickly

**Comment by ignatz** [comment]
Ok, https://github.com/trailbaseio/trailbase/commit/44e0bb6f846e6347125f679f13ee622a9baaa165 should have addressed the issue also for dev builds selectively bumping the optimization level. Let me know if it didn't fix your issue

**Comment by vootooguru** [comment]
Okay, I'll verify it tomorrow. Thanks for your support.

**Comment by vootooguru** [comment]
The verification is complete, and the version 44e0bb6 has resolved my issue. Thank you for the support.

> Ok, [44e0bb6](https://github.com/trailbaseio/trailbase/commit/44e0bb6f846e6347125f679f13ee622a9baaa165) should have addressed the issue also for dev builds selectively bumping the optimization level. Let me know if it didn't fix your issue
